tunnel/tests/integration/main.rs

```diff
@@ -314,10 +373,31 @@ fn assert_valid_refs(events: &[TracingEvent]) {
                 id, metadata_id, ..
             } => {
                 assert!(span_ids.insert(*id));
-                assert!(call_site_ids.contains(metadata_id));
+                if !call_site_ids.contains(metadata_id) {
+                    // Enhanced error reporting for potential race condition detection
+                    eprintln!("Event ordering issue detected at event {}: NewSpan references unknown metadata_id {}", i, metadata_id);
+                    eprintln!("Known call site IDs: {:?}", call_site_ids);
+                    eprintln!(
+                        "Event context (events {}..{}):",
+                        i.saturating_sub(3),
+                        (i + 3).min(events.len())
+                    );
+                    for (j, ctx_event) in events[i.saturating_sub(3)..(i + 3).min(events.len())]
+                        .iter()
+                        .enumerate()
+                    {
+                        let event_idx = i.saturating_sub(3) + j;
+                        let marker = if event_idx == i { " >>> " } else { "     " };
+                        eprintln!("{}[{}] {:?}", marker, event_idx, ctx_event);
+                    }
```

Comment:
Bikeshedding: IMO, that's too much effort for code that should never be triggered. It'd be sufficient to include `{events:#?}` in the panic message.